python/lsst/daf/butler/registry/collections/_base.py

```diff
@@ -408,31 +409,90 @@ def _fetch_by_key(self, collection_ids: Iterable[K] | None) -> list[CollectionRe
-            self._insert_collection_chain_rows(chain.key, 0, [child.key for child in child_records])
+            self._insert_collection_chain_rows(c.parent_key, 0, c.child_keys)
 
-        record = ChainedCollectionRecord[K](chain.key, chain.name, children=tuple(names))
+        names = [child.name for child in c.child_records]
+        record = ChainedCollectionRecord[K](c.parent_key, parent_collection_name, children=tuple(names))
         self._addCachedRecord(record)
-        return record
+
+    def prepend_collection_chain(
+        self, parent_collection_name: str, child_collection_names: list[str]
+    ) -> None:
+        self._add_to_collection_chain(
+            parent_collection_name, child_collection_names, self._find_prepend_position
+        )
+
+    def extend_collection_chain(self, parent_collection_name: str, child_collection_names: list[str]) -> None:
+        self._add_to_collection_chain(
+            parent_collection_name, child_collection_names, self._find_extend_position
+        )
+
+    def _add_to_collection_chain(
+        self,
+        parent_collection_name: str,
+        child_collection_names: list[str],
+        position_func: Callable[[_CollectionChainModificationContext], int],
+    ) -> None:
+        with self._modify_collection_chain(parent_collection_name, child_collection_names) as c:
+            self._remove_collection_chain_rows(c.parent_key, c.child_keys)
+            starting_position = position_func(c)
+            self._block_for_concurrency_test()
+            self._insert_collection_chain_rows(c.parent_key, starting_position, c.child_keys)
+
+    def remove_from_collection_chain(
+        self, parent_collection_name: str, child_collection_names: list[str]
+    ) -> None:
+        with self._modify_collection_chain(
+            parent_collection_name,
+            child_collection_names,
+            # Removing members from a chain can't create collection cycles
+            skip_cycle_check=True,
+        ) as c:
+            self._remove_collection_chain_rows(c.parent_key, c.child_keys)
+
+    @contextmanager
+    def _modify_collection_chain(
+        self,
+        parent_collection_name: str,
+        child_collection_names: list[str],
+        *,
+        skip_caching_check: bool = False,
+        skip_cycle_check: bool = False,
+    ) -> Iterator[_CollectionChainModificationContext[K]]:
+        if (not skip_caching_check) and self._caching_context.is_enabled:
+            # Avoid having cache-maintenance code around that is unlikely to
+            # ever be used.
+            raise RuntimeError("Chained collection modification not permitted with active caching context.")
+
+        if not skip_cycle_check:
+            self._sanity_check_collection_cycles(parent_collection_name, child_collection_names)
+
+        child_records = self.resolve_wildcard(
+            CollectionWildcard.from_names(child_collection_names), flatten_chains=False
+        )
```

Comment:
> It seems a little wasteful to query for child rows both here and in the sanity check, though I don't know if that's likely to matter.

It's definitely wasteful but I think not to an extent that is worth worrying about... in any case it's no slower than the previous version because all of this was re-used from the existing implementation.

The sanity check one is restricted to only `type==CHAINED` and is recursive so it's sort of a different query.  I think the way these two queries are done will change up some more when we add the recursive query thing to resolve_wildcard... all we need from the second query is the IDs corresponding to the names and it does more work than necessary right now.

> As is, couldn't another process modify the chain in between the queries for its contents and the point at which we acquire the lock? 

These aren't queries for the chain contents -- they're queries about the user-provided list of new child collections, and do not involve the parent collection.

The lock is only on the parent collection.  So moving these inside the lock wouldn't buy us anything and would make the critical section longer.  

For the child name -> ID query, the only concurrent change that affects anything is deleting one of the children (or delete and replace with a new collection of the same name).  In those cases the foreign key constraint will prevent us from inserting a child that existed at the time we looked up its ID but no longer exists.

The sanity check thing is hairier, and as I mentioned in the comment does not guarantee that cycles cannot exist.  To guarantee no cycles we'd have to recursively lock all child chained collections as well.  The locking order is  tricky to get right and locking multiple collections makes the availability issues with this approach worse.  It's isomorphic to the locking for the denormalized, flattened chain table we discussed the other week.  (That table would actually be a good way to enforce this... we could have a check constraint that `parent != child`.) But in any case just locking the parent doesn't help any -- this query wants to know if the parent is a recursive child of any of the children.  Locking the parent only prevents its children from changing and we don't care about the parent's children.

> I think some comments would help.
Yeah I'll add a bit more info here.